### ISSUE TITLE
Guidance on software engineering for data science

### ISSUE THREAD

**Issue by mloning** [alan-turing-institute/the-turing-way#1721]
### Summary
It seems that there is some guidance missing on software engineering/design for data science, based on a brief discussion on slack with @lauracion, @KirstieJane, @martinagvilas and @malvikasharan.

* There are lots of generic software engineering resources, but few are focused on data science, and it is often not obvious how to translate them into research code.
* Data scientists often lack a formal education in software engineering, and pick up the relevant skills when having to write code, often without having the time or resources to think more carefully about design principles. 
* There is a lack of hands-on tutorials that motivate and illustrate design choices

I have co-authored a recent [paper](https://arxiv.org/abs/2101.04938) on software design for ML toolboxes, but we believe more work is needed to make it more accessible and useful for others. 

Any feedback or ideas are very welcome!

**Comment by KirstieJane** [comment]
Following on from a point that @malvikasharan raised on slack. I feel like your paper is much more focused than "software engineering" for data science. I haven't read the paper, but knowing the work that you've been doing for the last few years I think the huge added value is in the thoughtful _design_ of ML toolboxes. I would LOVE for that to be featured in the Project Design section of the book!! Kind of a "what do we wish you had thought about before you started your new project" vibe :)

**Comment by mloning** [comment]
Agreed, it's more about API design on the code level rather than software engineering more broadly.
